cub/cub/device/dispatch/dispatch_merge.cuh

```diff
@@ -218,13 +218,8 @@ template <typename KeyIt1,
           typename PolicyHub = device_merge_policy_hub<value_t<KeyIt1>, value_t<ValueIt1>>>
 struct dispatch_t
 {
-  using key_t   = cub::detail::value_t<KeyIt1>;
-  using value_t = cub::detail::value_t<ValueIt1>;
-
   // Cannot check output iterators, since they could be discard iterators, which do not have the right value_type
-  static_assert(::cuda::std::is_same<cub::detail::value_t<KeyIt2>, key_t>::value, "");
-  static_assert(::cuda::std::is_same<cub::detail::value_t<ValueIt2>, value_t>::value, "");
-  static_assert(::cuda::std::__invokable<CompareOp, key_t, key_t>::value,
+  static_assert(::cuda::std::__invokable<CompareOp, cub::detail::value_t<KeyIt1>, cub::detail::value_t<KeyIt1>>::value,
```

Comment:
@bernhardmgruber as discussed in https://github.com/NVIDIA/cccl/pull/1817#discussion_r1684950617, this is a breaking change and should go into device code.

Reply:
You seem to have added the static assert in the device code but haven't removed it from the dispatch.

Reply:
We discussed this static assertion about the return type:
```c++
::cuda::std::is_convertible<typename ::cuda::std::__invoke_of<CompareOp, key_t, key_t>::type, bool>::value
```
which I have removed. The offending static assert here was not commented at in PR #1817.

I thought querying whether a comparison operator is invokable with a given set of arguments should be fine in host code. Only asking for the return type is problematic. My reasoning was that it's much more user friendly to have a compile error here if the user passed a wrong operator, than many stacks deeper in the kernel code.

Reply:
> Only asking for the return type is problematic. 

Any form of introspection of the signature is problematic, including:

> Introspecting the parameter type of operator() is only supported in device code.

Let's move the check to device code.